metadata/reportssql/testing_report_functions.sql

```diff
@@ -553,6 +577,85 @@ BEGIN
 END$$ 
 DELIMITER ;
 
+-- getPatientAgeInMonthsAtDate
+
+DROP FUNCTION IF EXISTS getPatientAgeInMonthsAtDate;
+
+DELIMITER $$
+CREATE FUNCTION getPatientAgeInMonthsAtDate(
+    p_patientId INT(11),
+    p_date DATE) RETURNS VARCHAR(50)
+    DETERMINISTIC
+BEGIN
+    DECLARE result VARCHAR(50);
+
+    SELECT 
+        timestampdiff(MONTH, p.birthdate, p_date) INTO result 
+    FROM person p 
+    WHERE p.voided = 0
+        AND p.person_id = p_patientId
+    LIMIT 1;
+
+    RETURN result;
+END$$
+DELIMITER ;
+
+-- getTestResultWithinReportingPeriod
+
+DROP FUNCTION IF EXISTS getTestResultWithinReportingPeriod;
+
+DELIMITER $$
+CREATE FUNCTION getTestResultWithinReportingPeriod(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE,
+    p_testUuid VARCHAR(38),
+    p_testDateUuid VARCHAR(38)) RETURNS VARCHAR(50)
+    DETERMINISTIC
+BEGIN
+    DECLARE result VARCHAR(50);
+
+    -- retrieve the test result from OpenElis
+    SELECT cn.name INTO result
+    FROM obs o
+        JOIN concept_name cn ON o.value_coded = cn.concept_id AND cn.locale = "en"
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND o.order_id IS NOT NULL
+        AND o.concept_id = (SELECT c.concept_id FROM concept c WHERE c.uuid = p_testUuid LIMIT 1)
+        AND o.date_created BETWEEN p_startDate AND p_endDate
+    LIMIT 1;
+
+    IF (result IS NOT NULL) THEN
+        RETURN result;
+    END IF;
+
+    -- retrieve the test result from the lab form
+    SELECT cn.name INTO result
+    FROM obs o
+        JOIN concept_name cn ON o.value_coded = cn.concept_id AND cn.locale = "en"
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND o.order_id IS NULL
+        AND o.concept_id = (SELECT c.concept_id FROM concept c WHERE c.uuid = p_testUuid LIMIT 1)
+        AND ( -- query the test date
+            SELECT o2.value_datetime
+            FROM obs o2
+            WHERE o2.encounter_id = o.encounter_id
+                AND o2.voided = 0
+                AND o2.concept_id = (
+                    SELECT c.concept_id
+                    FROM concept c
+                    WHERE c.uuid = p_testDateUuid
+                    LIMIT 1)
+            LIMIT 1
+        ) BETWEEN p_startDate AND p_endDate
```

Comment:
@tmvumbi2 Test case 11 failed. We might have to change the logic here to select the latest value_datetime?

Reply:
This test case is still failing. These are my test steps:

`Steps:

1- Start visit, add manual lab pcr Negative, test date 3 March

mysql> select getTestResultWithinReportingPeriod(358,'2020-03-01','2020-03-15',"a5239a85-6f75-4882-9b9b-60168e54b7da","9bb7b360-3790-4e1a-8aca-0d1341663040");
+-------------------------------------------------------------------------------------------------------------------------------------------------+
| getTestResultWithinReportingPeriod(358,'2020-03-01','2020-03-15',"a5239a85-6f75-4882-9b9b-60168e54b7da","9bb7b360-3790-4e1a-8aca-0d1341663040") |
+-------------------------------------------------------------------------------------------------------------------------------------------------+
| Negative                                                                                                                                        |
+-------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)


2- Close visit, start new one and add manual lab pcr Positive, test date 11 March

mysql> select getTestResultWithinReportingPeriod(358,'2020-03-01','2020-03-15',"a5239a85-6f75-4882-9b9b-60168e54b7da","9bb7b360-3790-4e1a-8aca-0d1341663040");
+-------------------------------------------------------------------------------------------------------------------------------------------------+
| getTestResultWithinReportingPeriod(358,'2020-03-01','2020-03-15',"a5239a85-6f75-4882-9b9b-60168e54b7da","9bb7b360-3790-4e1a-8aca-0d1341663040") |
+-------------------------------------------------------------------------------------------------------------------------------------------------+
| Negative                                                                                                                                        |
+-------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)`